add navigators

taylorli2000 opened the pull request:
SDCP-996

add navigators needed in https://github.com/superdesk/superdesk-cp/pull/474

taylorli2000 committed:
add navigators

taylorli2000 committed:
add styles

tomaskikutis commented on app-typescript/components/DatePicker.tsx:
```diff
@@ -13,7 +13,13 @@ import {getMonthNames, getWeekdayNames} from '@sourcefabric/common';
 
-interface IDatePickerBase extends IInputWrapper {
+interface NavigatorsDateRange {
+    showNavigators?: boolean;
```
Don't expose these as option - show navigators unconditionally.

tomaskikutis commented on app/styles/components/_datepicker.scss:
```diff
@@ -0,0 +1,15 @@
+.datepicker-calendar-panel {
+  div {
+    div {
+      div:nth-child(3) {
```
Use cleaner selectors. No need to count children or use nesting. Primereact exposes everything needed as classnames

<img width="1727" height="774" alt="image" src="https://github.com/user-attachments/assets/28926d94-5610-4da2-9c65-087c64499585" />

tomaskikutis reviewed (commented):
Adjust styles to maintain existing design:

<img width="474" height="480" alt="image" src="https://github.com/user-attachments/assets/2f3cbc75-8340-4d8a-adc0-541022ee7869" />

taylorli2000 committed:
fix styles, remove showNavigators prop

tomaskikutis commented on app-typescript/components/DatePicker.tsx:
```diff
@@ -176,6 +179,15 @@ export class DatePicker extends React.PureComponent<IDatePicker, IState> {
 
         const showClearButton = this.props.required === true ? false : this.props.hideClearButton !== true;
 
+        const navigatorsProps = {
```
Iinline these props where they're used. It's only used once thus there is no use to move it into a constant.

tomaskikutis reviewed (commented):
<img width="323" height="326" alt="image" src="https://github.com/user-attachments/assets/a8d412e7-05dc-478f-a704-5703be40acc8" />

There's an issue with caret alignment when short month names are used. That's because native select doesn't support having width same as currently selected value.

I've released `@superdesk/primereact@5.0.2-16` where [I've implemented passing custom select components for month/year](https://github.com/superdesk/primereact/commit/23de78c886589cb8da85120c1aa407bf8eec2741).

Use it pass a select component implementation that doesn't have the issue with caret alignment. It should work to configure `TreeSelect` for this - it supports custom templates.

tomaskikutis commented:
<img width="1086" height="528" alt="image" src="https://github.com/user-attachments/assets/841481f2-fb6c-4fe9-afad-85f8ff1083f8" />

Also use this font icon instead of adding a new image file. I understand it probably wasn't possible with native select - but will be possible when using `TreeSelect`

taylorli2000 committed:
use tree select, add clearable prop for remove button

taylorli2000 commented:
> <img alt="image" width="323" height="326" src="https://private-user-images.githubusercontent.com/2076953/500581699-a8d412e7-05dc-478f-a704-5703be40acc8.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NjA0NjYxNTcsIm5iZiI6MTc2MDQ2NTg1NywicGF0aCI6Ii8yMDc2OTUzLzUwMDU4MTY5OS1hOGQ0MTJlNy0wNWRjLTQ3OGYtYTcwNC01NzAzYmU0MGFjYzgucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MTAxNCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTEwMTRUMTgxNzM3WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ZGJkNjFiYjZhMTIyOTY1NWU1MWM5YjU0ZDQzZTViZGFmNjVhNDRiODc5NmJkNDFiMDk5ODg0NWFiMTk2NzdjOSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.s8BierXDR7oc5mBWypT2ZdIn__oLS91Gv9X0m52sZqM">
> There's an issue with caret alignment when short month names are used. That's because native select doesn't support having width same as currently selected value.
> 
> I've released `@superdesk/primereact@5.0.2-16` where [I've implemented passing custom select components for month/year](https://github.com/superdesk/primereact/commit/23de78c886589cb8da85120c1aa407bf8eec2741).
> 
> Use it pass a select component implementation that doesn't have the issue with caret alignment. It should work to configure `TreeSelect` for this - it supports custom templates.

I've updated prime react to pass in the tree selects as custom selects. I added a clearable prop to hide the remove button since required and readonly don't fit here. carets are now aligned as well and styled accordingly

tomaskikutis commented on app-typescript/components/DatePicker.tsx:
```diff
@@ -334,3 +334,41 @@ export class DatePickerISO extends React.PureComponent<IDatePickerISO> {
+    <TreeSelect<string>
+        kind="synchronous"
+        value={[value]}
+        getOptions={() => options.map((o) => ({...o, value: o.label}))}
```
Do not use naming shorhands - use `option` instead of `o`.

Another issue is that your return value doesn't match the interface. Due to use of spread operator typescript isn't flagging it. That's a good reason to avoid spreads.

tomaskikutis replied:
I see you still haven't addressed types not matching. Drop spread operator and you'll see the typescript issue

tomaskikutis reviewed (commented):
https://github.com/user-attachments/assets/9c6f174d-2c7d-457d-8988-5cfdd18b96d3

- [x] fix date picker closing after selecting a month/year
- [x] improve CSS code quality - only add styles for classnames that you define - use `valueTemplate` prop of tree select to pass a custom template.

tomaskikutis commented:
Also update styles to match the screenshot https://github.com/superdesk/superdesk-ui-framework/pull/969#pullrequestreview-3319761730 Currently font colors/weight and spacing are not matching. Use CSS existing CSS variables to find similar colors for text.

taylorli2000 committed:
stop bubbling on navigator, add value and wrapper templates

tomaskikutis commented on app-typescript/components/TreeSelect/TreeSelect.tsx:
```diff
@@ -273,6 +274,11 @@ export class TreeSelect<T> extends React.Component<IProps<T>, IState<T>> {
                 }
             });
 
+            this.dropdownRef.current?.addEventListener('mousedown', (e: MouseEvent) => {
```
Is this one to prevent date picker closing after selecting a month/year? If so the code should go to date picker - not tree select.

tomaskikutis commented on app/styles/primereact/_pr-datepicker.scss:
```diff
@@ -9,13 +9,80 @@
+        }
+
+        .sd-datepicker__navigator-value {
+            z-index: -1;
```
z index isn't needed here is it?

tomaskikutis commented on app/styles/primereact/_pr-datepicker.scss:
```diff
@@ -9,13 +9,80 @@
+        gap: $sd-base-increment * 0.2;
+
+        .sd-datepicker__navigator-wrapper {
+            * {
```
avoid star operator, list all elements you are targetting explicitly

tomaskikutis committed:
revert formatting; drop select backgrounds

tomaskikutis committed:
drop unused code

taylorli2000 committed:
add dropdown click callback, fix getOptions, styles

tomaskikutis committed:
improve code that prevents popover from closing

tomaskikutis committed:
improve styles

tomaskikutis committed:
4.1.2

tomaskikutis committed:
Merge branch 'v4' into add-navigators

tomaskikutis committed:
fix lint
